tensorboard/data/server/tf_record.rs

```
@@ -0,0 +1,327 @@
+            let actual_crc = MaskedCrc::compute(length_buf);
+            if length_crc != actual_crc {
+                return Err(ReadRecordError::BadLengthCrc(ChecksumError {
+                    got: actual_crc,
+                    want: length_crc,
+                }));
+            }
+
+            let length = LittleEndian::read_u64(length_buf);
+            let data_plus_footer_length_u64 = length + (FOOTER_LENGTH as u64);
+            let data_plus_footer_length = data_plus_footer_length_u64 as usize;
+            if data_plus_footer_length as u64 != data_plus_footer_length_u64 {
+                return Err(ReadRecordError::TooLarge(length));
+            }
+            self.data_plus_footer.reserve_exact(data_plus_footer_length);
+        }
+
+        if self.data_plus_footer.len() < self.data_plus_footer.capacity() {
+            read_remaining(reader, &mut self.data_plus_footer)?;
+        }
+
+        let data_length = self.data_plus_footer.len() - FOOTER_LENGTH;
+        let data_crc_buf = self.data_plus_footer.split_off(data_length);
+        let data = std::mem::take(&mut self.data_plus_footer);
+        let data_crc = MaskedCrc(LittleEndian::read_u32(&data_crc_buf));
+        self.header.clear(); // reset; caller may use this again
+        Ok(TfRecord { data, data_crc })
+    }
+}
+
+/// Fills `buf`'s remaining capacity from `reader`, or fails with `Truncated` if the reader is dry.
+fn read_remaining<R: Read>(reader: &mut R, buf: &mut Vec<u8>) -> Result<(), ReadRecordError> {
+    let want = buf.capacity() - buf.len();
+    reader.take(want as u64).read_to_end(buf)?;
+    if buf.len() < buf.capacity() {
+        return Err(ReadRecordError::Truncated);
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::scripted_reader::ScriptedReader;
+    use std::io::Cursor;
+
+    #[test]
+    /// Tests a happy path with multiple records, one of which is truncated.
+    fn test_success() {
```

Is it hypothetically possible that the writer is so borked that it writes more data than alleged length? If so, what would be a possible recovery mechanism? I cannot think of one so answer can be, writer **should never** be borked and, when it is, we will fail hard.

---

Yup, that's what I was asking. Thanks for the response.

I assumed you can craft the data so that it contains correct "header" with very large length (with matching length CRC). In this case, since we forbid allocating memory larger than 2 GiB, the impact is minimal and will not cause our program to allocate 64 GiB and will not cause resource starvation.  LGTM!